opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -1686,18 +1687,17 @@ namespace Opm {
                 const int w = well->indexOfWell();
                 if (!well->isOperable() ) continue;
 
-                if (this->wellState().effectiveEventsOccurred(w) ) {
+                auto& events = this->wellState().events();
+                if (events.hasEvent(well->name(), WellStateFullyImplicitBlackoil::event_mask)) {
                     well->updateWellStateWithTarget(ebosSimulator_, this->wellState(), deferred_logger);
-                }
 
-                // there is no new well control change input within a report step,
-                // so next time step, the well does not consider to have effective events anymore
-                // TODO: if we can know whether this is the first time step within the report step,
-                // we do not need to set it to false
-                // TODO: we should do this at the end of the time step in case we will need it within
-                // this time step somewhere
-                if (this->wellState().effectiveEventsOccurred(w) ) {
-                    this->wellState().setEffectiveEventsOccurred(w, false);
+                    // there is no new well control change input within a report step,
+                    // so next time step, the well does not consider to have effective events anymore
+                    // TODO: if we can know whether this is the first time step within the report step,
+                    // we do not need to set it to false
+                    // TODO: we should do this at the end of the time step in case we will need it within
+                    // this time step somewhere
+                    events.clearEvent(well->name(), WellStateFullyImplicitBlackoil::event_mask);
                 }
```

Comment:
We need to be a bit careful here. The events may be used by other part of the code.

Reply:
Well - og the tests pass the tests pass - or?

Reply:
> Well - if the tests pass the tests pass - or?

I guess so. I like your change of making it mutable. That would make it possible to add an event when a well is opened during the simulations. For instance due to WTEST or some ACTION keyword. I have looked through the code and adding the clear logic here is ok. That would assure that updateWellStateWithTarget is only triggered once for each new event. 

I would suggest that we remove the TODO comments and move the clearEvent to just after the updateWellStateWithTarget to make it clear why we want to clear the event. I have a PR #3176 that touch the same code. That would need a rebase after this is merged. I can do the clean-up there.

Reply:
OK - I have removed the comment in a fixup. NB: This comment was not originally written by me - it was green in the diff only because of some minor reformatting.

Reply:
>  NB: This comment was not originally written by me

I know. So thanks for the cleanup. I must admit I have never understood the logic of he setEffectiveEventsOccurred(w,false) before I read carefully through it today. I think the code is more readable after your fixup commit. You could consider adding a comment like
"We only want update the wellstate with the targets the first time the event occurs." 
before the events.clearEvent(...)
if you plan to do a squash anyway.

Reply:
OK - done.